libs/eavmlib/src/ahttp_client.erl

```diff
@@ -0,0 +1,443 @@
+%%          be identified from the same reference.
+%% @end
+%%-----------------------------------------------------------------------------
+-spec request(
+    Conn :: connection(),
+    Method :: iodata(),
+    Path :: iodata(),
+    Headers :: [iodata()],
+    Body :: binary() | undefined | nil | stream
+) -> {ok, connection(), reference()} | error_tuple().
+request(
+    #http_client{host = Host, socket = Socket} = Conn, Method, Path, Headers, Body
+) ->
+    {BodyBin, MaybeBodyHeaders} = prepare_body(Body, Headers),
+    HeadersList = transform_headers(MaybeBodyHeaders),
+    Data = iolist_to_binary([
+        Method,
+        $\s,
+        Path,
+        <<" HTTP/1.1\r\n">>,
+        <<"Host: ">>,
+        Host,
+        <<"\r\n">>,
+        HeadersList,
+        <<"\r\n">>,
+        BodyBin
+    ]),
+    case send(Socket, Data) of
+        ok ->
+            Ref = make_ref(),
+            WantedHeaders = ?DEFAULT_WANTED_HEADERS ++ Conn#http_client.parse_headers,
+            {ok,
+                Conn#http_client{ref = Ref, parser = #parser_state{wanted_headers = WantedHeaders}},
+                Ref};
+        {error, _} = Error ->
+            Error
+    end.
+
+send({gen_tcp, TCPSocket}, Data) ->
+    case gen_tcp:send(TCPSocket, Data) of
+        ok -> ok;
+        {error, Error} -> {error, {gen_tcp, Error}}
+    end;
+send({ssl, SSLSocket}, Data) ->
+    case ssl:send(SSLSocket, Data) of
+        ok -> ok;
+        {error, Error} -> {error, {ssl, Error}}
+    end.
+
+prepare_body(undefined, Headers) ->
+    {<<"">>, Headers};
+prepare_body(nil, Headers) ->
+    {<<"">>, Headers};
+prepare_body(stream, Headers) ->
+    {<<"">>, Headers};
+prepare_body(Body, Headers) ->
+    BodyLen = integer_to_binary(byte_size(Body)),
+    {Body, [{<<"content-length">>, BodyLen} | Headers]}.
+
+transform_headers([]) ->
+    [];
+transform_headers([{Name, Value} | Tail]) ->
+    [Name, <<": ">>, Value, <<"\r\n">> | transform_headers(Tail)].
+
+%%-----------------------------------------------------------------------------
+%% @param   Conn the connection
+%% @param   Msg a received message
+%% @returns Either a list of responses, `unknown' or an error tuple.
+%% @doc     This function should be used when in active mode in order to process socket messages.
+%%
+%%          If a socket message is streamed using this function, a tuple with a list of http
+%%          responses is returned (e.g. `{ok, UpdatedConn, Responses}`', or an error tuple.
+%%
+%%          Otherwise `unknown' is returned, that means that the message is not a socket message
+%%          tied to the open connection, and it should be handled in some other way.
+%%
+%%          The first returned response to a new request is a status `{status, Ref, 200}' for
+%%          a successful response. After that headers and data may follow.
+%%
+%%          Since the response might span multiple socket messages, `stream/3' may be called
```

Comment:
Should this be `stream/2`? I don't see a `stream/3`.

Reply:
Fixed.